Gems/Terrain/Assets/Shaders/Terrain/TerrainMacroTextureClipmapGenerationPass.azsl

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+#include "TerrainMacroTextureClipmapPassSrg.azsli"
+#include "ClipmapComputeHelpers.azsli"
+#include <Atom/RPI/TangentSpace.azsli>
+
+void SampleTexture(uint2 pixelPosition, uint clipmapLevel, out float3 macroColor, out float3 macroNormal)
+{
+    float2 worldPosition = GetCurrentWorldPosition(pixelPosition, clipmapLevel);
+    
+    uint2 macroGridResolution = uint2(TerrainSrg::m_macroMaterialGrid.m_resolution >> 16, TerrainSrg::m_macroMaterialGrid.m_resolution & 0xFFFF);
```

Comment:
Many of the operations like this one and a few others I think are actually wave and shader uniform and probably best computed on the CPU and passed to the shader as data.

Reply:
The sampling code most comes from existing code used in forward pass. We have a plan to change the macro texture representation into quad tree instead of dense tiles, so for now I'll leave as it is and keep a note when working upgrading the data structure.